test/test_broker.hpp

```diff
@@ -825,8 +839,31 @@ class test_broker {
         MQTT_NS::buffer contents,
         MQTT_NS::publish_options pubopts,
         MQTT_NS::v5::properties props) {
+
+        std::map< con_sp_t, MQTT_NS::qos > subscribers;
```

Comment:
There's no need to construct a new std::map object.

Just put the body of your for-loop below directly into the lambda that you're passing to the find function.

Reply:
Map is needed to determine the max qos for all subscriptions. It is possible a topic matches multiple subscriptions from a client, but the message needs to be published only once.

Reply:
Actually, I'm not sure this is true.

https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901120

> When Clients make subscriptions with Topic Filters that include wildcards, it is possible for a Client’s subscriptions to overlap so that a published message might match multiple filters. In this case the Server MUST deliver the message to the Client respecting the maximum QoS of all the matching subscriptions [MQTT-3.3.4-2]. In addition, the Server MAY deliver further copies of the message, one for each additional matching subscription and respecting the subscription’s QoS in each case.


One way to implement this is to publish the message once for each subscription that matches. Another way is to publish the message once total, with the maximum QOS of the subscriptions that match.

Honestly, I don't think this is the right layer to be accomplishing this task. Publishing a message with mqtt_cpp is cheap, since the data doesn't actually get copied (if you're using the zero-copy api, anyway), and for a connection that's had a while to initialize, all of the processing can be done without allocations (e.g. re-use space from previous allocations).

But with the assumption that we want to only sent the publish message one time, with the maximum QOS, then there are two ways to accomplish that.

1) Do what you're doing here, which requires an allocation for each endpoint, not so amazing, especially since it's happening on the stack, so there's no hope of the process settling eventually into an allocationless state.
2) Have some mechanism in the endpoint.hpp code to allow a publish message to inspect the list of pending packets, and if the new packet matches (with some predicate passed in during send, probably), then update in place instead of sending new.

However, approach 2 isn't so great in a multithreaded context, so instead having some kind of staging area for outgoing packets might be nice, that would allow each endpoint to have their own staging area, which means the allocations would eventually calm down and you'd settle into an allocation-less path.

An easy way to approach that would be to extend endpoint.hpp with another layer of derived code just for the broker, so that the endpoint being created by the socket acceptor code could allocate space for the vector of packets to be sent out.

Then this for-loop could save itself the trouble of allocating the map object.

Reply:
Unrelated to what I said above:

What's the max size that this map will grow to?

Reply:
The max size of the map is possibly all connections which are active.

Reply:
Maybe the result should be sorted based on the connection, this way you can determine the max qos per connection quite easily on the stack. I will think about what would be the best approach for this.

Reply:
Also note that the subscription matching now also does heap allocations, but this could be avoided by using a depth-first search. But this would involve limiting the maximum number of levels we support, otherwise the stack may overflow.

Reply:
You could do some kind of staging in the endpoint, because you would publish the same message multiple times during staging. This would not require extra allocation, you "stage" the message in the connection and calculate the max qos. This could be done in the broker itself per connection or in the endpoint.

Reply:
Actually, you don't need to stage the message content, only the qos, so we could add a staged_qos value to the connection structure.

Reply:
Right. the endpoint class is intended to be inherited from (see callable_overlay for example on how that's done)

So an additional overlay / derived type could be inserted into the type hierarchy and the per-endpoint storage needed for this operation could be set aside there.

Reply:
The active sessions structure could also work for this.

However. The test broker doesn't play carefully with multiple threads. 

Please consider the case of two threads processing an incoming publish for which the same endpoint has subscribed to both.

Without some synchronization between the threads you'll overwrite your staged QoS value.

Not saying you need to solve that. Since scaling to multiple threads won't work properly right now anyway. But its something to keep in mind.

Reply:
I would say it has to be part of session, you publish a message to a session and store it in the session if the connection is offline.

Reply:
Yes, well, regarding the multithreading, i would propose a different architecture anyway. Indeed many of the datastructures are not thread-safe now. But discussing multi-threading broker architecture design is possibly a whole other discussion.